### PR TITLE
Booking additional fields

### PR DESCRIPTION
**Pull request by folix-01**
You can also create the addtional fields for your booking, you just need to compile 
them in your PrenotazioneType.
And they will appear in the ["booking_types"]["booking_additional_fields_schema"] 
in your booking schema so u can compile them for your booking in this way:

### CODE REVIEW

**Comment by mamico** [comment]
README
```
You can also create the addtional fields for your booking, you just need to compile them in your PrenotazioneType.
And they will appear in the ["booking_types"]["booking_additional_fields_schema"] in your booking schema so u can compile them for your booking in this way:
```

**Comment by mamico** [comment]
How does the frontend know about the additional fields? By reading the typology object? 

Alternatively, if you put the additional fields in the schema https://github.com/RedTurtle/redturtle.prenotazioni/blob/main/src/redturtle/prenotazioni/restapi/services/booking_schema/get.py the only change on the frontend would be to call the schema api with the chosen typology.

At that point even the scripting API would have no changes on the frontend and it would be the backend that would save the additional fields to the correct field.

\cc @folix-01 @giuliaghisini

**Comment by folix-01** [comment]
> @folix-01 added a test for patch method that fail with `additional_fields` [416f9b7](https://github.com/RedTurtle/redturtle.prenotazioni/commit/416f9b700c5e7d1d24b8aa660127075212b56b91)

Do we use the edit on ioprenoto?

**Comment by mamico** [comment]
> > @folix-01 added a test for patch method that fail with `additional_fields` [416f9b7](https://github.com/RedTurtle/redturtle.prenotazioni/commit/416f9b700c5e7d1d24b8aa660127075212b56b91)
> 
> Do we use the edit on ioprenoto?

yep

File: README.rst [hunk 1]
```diff
@@ -510,6 +510,7 @@ Response::
                 {
                 "duration": "60",
                 "name": "Rilascio CIE"
+                "booking_additional_fields_schema": {"name": "field1", "desctiption": "Field number 1", "type": "text"}
```
Comment (cekk) [inline README.rst:513]: typo

File: README.rst [hunk 2]
```diff
@@ -703,6 +704,66 @@ Response::
         ]
     }
 
+Booking Additional Fields
+=========================
+
+You can also create the addtional fields for your booking, you just need to compile 
+them in your PrenotazioneType.
+And they will appear in the ["booking_types"]["booking_additional_fields_schema"] 
+in your booking schema so u can compile them for your booking in this way:
```
Comment (cekk) [inline README.rst:713]:
```suggestion
in your booking schema so you can compile them for your booking in this way:
```

File: README.rst [hunk 3]
```diff
@@ -703,6 +704,66 @@ Response::
...
+
+Example::
+
+    curl http://localhost:8080/Plone/++api++/<booking_folder_path>/@booking \
+        -X POST \
+        -H 'Accept: application/json' \
+        -H 'Content-Type: application/json' \
+        -d '{
+            "booking_date": "2023-05-23T09:00:00+02:00",
+            "booking_type": "SPID: SOLO riconoscimento \"de visu\" (no registrazione)",
```
Comment (cekk) [inline README.rst:733]:
```suggestion
            "booking_type": "Type x",
```

File: README.rst [hunk 4]
```diff
@@ -703,6 +704,66 @@ Response::
...
+            "additional_fields": [{"name": "field1", "value": "Addional field text"}]
+        }'
+
+Response::
+
+    {
+        "booking_code": "17E3E6",
+        "booking_date": "2023-05-22T09:09:00",
+        "booking_expiration_date": "2023-05-22T09:10:00",
+        "booking_type": "SPID: SOLO riconoscimento \"de visu\" (no registrazione)",
```
Comment (cekk) [inline README.rst:747]:
```suggestion
        "booking_type": "Type x",
```

File: README.rst [hunk 5]
```diff
@@ -703,6 +704,66 @@ Response::
...
+        "company": null,
+        "cosa_serve": null,
+        "description": "",
+        "email": "mario.rossi@example",
+        "fiscalcode": "",
+        "gate": "postazione1",
```
Comment (cekk) [inline README.rst:753]:
```suggestion
        "gate": "gate 1",
```

File: src/redturtle/prenotazioni/adapters/booker.py [hunk 6]
```diff
@@ -191,6 +192,11 @@ def generate_params(self, data, force_gate, duration):
         if fiscalcode:
             params["fiscalcode"] = fiscalcode.upper()
 
+        additional_fields = data.get("additional_fields", "")
+
+        if additional_fields:
+            params["additional_fields"] = json.dumps(additional_fields)
```
Comment (cekk) [inline src/redturtle/prenotazioni/adapters/booker.py:198]: why not using *json_compatible* method from plone.restapi?

File: src/redturtle/prenotazioni/content/prenotazione.py [hunk 7]
```diff
@@ -186,6 +186,13 @@ class IPrenotazione(model.Schema):
         required=False, title=_("label_booking_staff_notes", "Staff notes")
     )
 
+    # Schema is defined in PrenotaizioneType ad an datagridfield, and here we save the data as an json
```
Comment (cekk) [inline src/redturtle/prenotazioni/content/prenotazione.py:189]: typo

File: src/redturtle/prenotazioni/content/prenotazione.py [hunk 8]
```diff
@@ -186,6 +186,13 @@ class IPrenotazione(model.Schema):
...
+    # in base of selected type
+    additional_fields = schema.TextLine(
```
Comment (cekk) [inline src/redturtle/prenotazioni/content/prenotazione.py:191]: why not using a JSONfield like blocks one?

File: src/redturtle/prenotazioni/restapi/serializers/adapters/prenotazione.py [hunk 9]
```diff
@@ -100,6 +100,9 @@ def __call__(self, *args, **kwargs):
             "requirements": requirements,
             "modification_date": json_compatible(self.prenotazione.modified()),
             "creation_date": json_compatible(self.prenotazione.created()),
+            "additional_fields": json.loads(
```
Comment (cekk) [inline src/redturtle/prenotazioni/restapi/serializers/adapters/prenotazione.py:103]: if you use a JSONfield in prenotazioneType, you don't need this probably

File: src/redturtle/prenotazioni/restapi/serializers/adapters/prenotazione.py [hunk 10]
```diff
@@ -134,6 +137,9 @@ def __call__(self, *args, **kwargs):
             "email": self.prenotazione.email,
             "fiscalcode": self.prenotazione.fiscalcode,
             "phone": self.prenotazione.phone,
+            "additional_fields": json.loads(
```
Comment (cekk) [inline src/redturtle/prenotazioni/restapi/serializers/adapters/prenotazione.py:140]: same here

File: src/redturtle/prenotazioni/restapi/services/booking/update.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+class UpdateBooking(ContentPatch):
+    def reply(self, *args, **kwargs):
+
+        self.handle_additional_fields()
+
+        return super().reply(*args, **kwargs)
+
+    def handle_additional_fields(self):
+        data = json.loads(self.request.get("BODY", b"").decode())
```
Comment (cekk) [inline src/redturtle/prenotazioni/restapi/services/booking/update.py:22]: there is an utility in plone.restapi to get form data

File: src/redturtle/prenotazioni/restapi/services/booking/update.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        booking_folder = self.context.getPrenotazioniFolder()
+
+        if data:
+            # booking.additional_fields validation below
+            additional_fields = data.get("additional_fields")
+
+            if additional_fields and type(additional_fields) is list:
```
Comment (cekk) [inline src/redturtle/prenotazioni/restapi/services/booking/update.py:29]: this validation seems the same as in add.py..can we manage only once? Maybe in a common class

**Review by cekk — CHANGES_REQUESTED** [review]

File: src/redturtle/prenotazioni/restapi/services/booking/update.py (hunk 12 above)
Comment (folix-01) [reply src/redturtle/prenotazioni/restapi/services/booking/update.py:29]: it's not the same

File: src/redturtle/prenotazioni/content/prenotazione.py (hunk 8 above)
Comment (folix-01) [reply src/redturtle/prenotazioni/content/prenotazione.py]: done

File: src/redturtle/prenotazioni/restapi/serializers/adapters/prenotazione.py (hunk 9 above)
Comment (folix-01) [reply src/redturtle/prenotazioni/restapi/serializers/adapters/prenotazione.py]: done

File: src/redturtle/prenotazioni/restapi/services/booking/update.py (hunk 11 above)
Comment (folix-01) [reply src/redturtle/prenotazioni/restapi/services/booking/update.py]: done

File: src/redturtle/prenotazioni/restapi/serializers/adapters/prenotazione.py (hunk 10 above)
Comment (folix-01) [reply src/redturtle/prenotazioni/restapi/serializers/adapters/prenotazione.py]: done

**Pull request merged**